esi_leap/db/sqlalchemy/api.py

```diff
@@ -334,7 +350,93 @@ def lease_destroy(lease_uuid):
         session.flush()
 
 
-def resource_verify_availability(r_type, r_uuid, start, end):
+# Owner Changes
+def owner_change_get_by_uuid(owner_change_uuid):
+    query = model_query(models.OwnerChange)
+    result = query.filter_by(uuid=owner_change_uuid).one_or_none()
+    return result
+
+
+def owner_change_get_all(filters):
+    query = model_query(models.OwnerChange)
+
+    start = filters.pop('start_time', None)
+    end = filters.pop('end_time', None)
+    status = filters.pop('status', None)
+    from_or_to_owner_id = filters.pop('from_or_to_owner_id', None)
+
+    query = query.filter_by(**filters)
+
+    if status:
+        query = query.filter((models.OwnerChange.status.in_(status)))
+
+    if start and end:
+        query = query.filter((start >= models.OwnerChange.start_time) &
+                             (end <= models.OwnerChange.end_time))
+
+    if from_or_to_owner_id:
+        query = query.filter(
+            (from_or_to_owner_id == models.OwnerChange.from_owner_id) |
+            (from_or_to_owner_id == models.OwnerChange.to_owner_id))
+
+    return query
+
+
+def owner_change_create(values):
+    owner_change_ref = models.OwnerChange()
+    owner_change_ref.update(values)
+
+    with _session_for_write() as session:
+        session.add(owner_change_ref)
+        session.flush()
+        return owner_change_ref
+
+
+def owner_change_update(owner_change_uuid, values):
+    with _session_for_write() as session:
+        query = model_query(models.OwnerChange)
+        owner_change_ref = query.filter_by(
+            uuid=owner_change_uuid).one_or_none()
+
+        values.pop('uuid', None)
+        values.pop('from_owner_id', None)
+        values.pop('to_owner_id', None)
+        values.pop('resource_type', None)
+        values.pop('resource_uuid', None)
+
+        start = values.get('start_time', None)
+        end = values.get('end_time', None)
+        if start is None:
+            start = owner_change_ref.start_time
+        if end is None:
+            end = owner_change_ref.end_time
+        if start >= end:
+            raise exception.InvalidTimeRange(resource="an owner_change",
+                                             start_time=str(start),
+                                             end_time=str(end))
+
+        owner_change_ref.update(values)
+        session.flush()
+        return owner_change_ref
+
+
+def owner_change_destroy(owner_change_uuid):
+    with _session_for_write() as session:
+
+        query = model_query(models.OwnerChange)
+        owner_change_ref = query.filter_by(
+            uuid=owner_change_uuid).one_or_none()
+
+        if not owner_change_ref:
+            raise exception.OwnerChangeNotFound(
+                owner_change_uuid=owner_change_uuid)
+        query.delete()
+        session.flush()
+
+
+# Resources
+def resource_verify_availability(r_type, r_uuid, start, end,
+                                 is_owner_change=False):
```

Comment:
is it necessary to do these offer and lease checks when verifying a time for an owner_change object?

Reply:
this is the one part that is a bit tricky to wrap my head around

Reply:
Yeah, this one was extremely difficult for me to wrap my head around as well. My reasoning here is that if owner A creates a lease or an offer for a time period, it's illogical to allow them to move ownership of the node to owner B, as it's unclear whether owner B should have the right to cancel those leases/offers, or even to see them.